Enable IAM

emlaver opened the pull request:
## What

Enable IAM and verify that RA and CDT tests are passing with IAM API Key.

## How

- Revert @tomblench pull request (#370) to add IAM back into replication
- Added IAM key support to: `pushReplicationTarget`, `pullReplicationSource`, `pushReplicationWithTarget`, `pullReplicationWithSource`
- Replication plist: Added `TEST_COUCH_IAM_API_KEY`
- @ricellis added IAM documentation

@tomblench commmon code commit:
- Move pull/push helpers into
  CloudantReplicationBase
- Remove custom setup/teardown code from
  Attachments
- Use push/pull helpers in Attachment, which are
  IAM-aware

## Testing
Added IAM tests `iosIamRAT` and `macosIamRAT` to RA axes in Jenkinsfile.

WIP: `testRemoteLastSequenceValueAfterPushReplication` failing with two errors for iOS and macOS:
```
((localLastSequence) != nil) failed
((localLastSequence[@"source_last_seq"]) equal to (jsonResponse[@"source_last_seq"])) failed: ("(null)") is not equal to ("3005")
```

ricellis commented on CDTDatastoreReplicationAcceptanceTests/CloudantReplicationBase.m:
```diff
@@ -107,6 +111,26 @@ - (NSString*)createTemporaryDirectoryAndReturnPath
+/**
+ Get the IAM access token.  Used for CRUD helper methods.
+ */
+-(NSString *) getIAMBearerToken {
```
I'm guessing there is no easy way to just call the method we have in the interceptor to do this instead of having this UNIRest based version of getting a token?

emlaver replied:
If we change the currently private `getBearerToken` to public, then we can use the interceptor to do this:
https://github.com/cloudant/CDTDatastore/blob/0cb4502e09ce5a801c580d8112510ff7e8c6cda0/CDTDatastore/HTTP/CDTIAMSessionCookieInterceptor.m#L97

ricellis replied:
Fair enough, I don't think it is worth exposing that for this purpose.

ricellis commented on CDTDatastoreReplicationAcceptanceTests/ReplicationAcceptance.m:
```diff
@@ -1957,6 +1982,69 @@ -(void) testURLConnectionChangeTrackerWithRealRemote
+            XCTFail(@"Should not be called");
+        };
+        
+        //NSURL *url = [self sharedDemoURL];
```
Delete this if it isn't needed

emlaver replied:
Removed in 21fc79e.

ricellis commented on CDTDatastoreReplicationAcceptanceTests/ReplicationAcceptance.m:
```diff
@@ -1957,6 +1982,69 @@ -(void) testURLConnectionChangeTrackerWithRealRemote
     XCTAssertTrue(changeTrackerGotChanges);
 }
 
+-(void) testURLConnectionChangeTrackerWithRealRemoteAndIAMKey
```
What is the particular motivation for this new test, is this code not covered by the RAT?

emlaver replied:
This was added to use `CDTIAMSessionCookieInterceptor` as part of the session in the `TDChangeTracker`.   
The test can be removed if we create `shared_todo_sample` database in the clientlibs-test account and add logic for creating an IAM interceptor in the [original test](https://github.com/cloudant/CDTDatastore/blob/734167816daa208bde38a02796d9dea3b68d24fa/CDTDatastoreReplicationAcceptanceTests/ReplicationAcceptance.m#L1903).

emlaver replied:
@ricellis Are you OK with keeping this test? [Updated link](https://github.com/cloudant/CDTDatastore/pull/388/commits/321959d670eb3a464b87825c542df553f66933a7#diff-172fa1574ebaa7fcd57f7680a6e18c26R1985) since I've rebased.

ricellis commented on Jenkinsfile:
```diff
@@ -101,7 +138,7 @@ stage('BuildAndTest') {
             }]
     // Add replication acceptance tests for the master branch
-    if (env.BRANCH_NAME == "master") {
+    // if (env.BRANCH_NAME == "master") {
```
Don't forget to remove this temp change :)

ricellis commented on Rakefile:
```diff
@@ -15,7 +15,7 @@ SAMPLE_IOS = "Project"
 
 # Destinations
-IPHONE_DEST = (ENV["IPHONE_DEST"] == nil || ENV["IPHONE_DEST"] == "null") ? 'platform=iOS Simulator,OS=latest,name=iPhone 5' : ENV["IPHONE_DEST"]
+IPHONE_DEST = (ENV["IPHONE_DEST"] == nil || ENV["IPHONE_DEST"] == "null") ? 'platform=iOS Simulator,OS=latest,name=iPhone SE' : ENV["IPHONE_DEST"]
```
Out of curiosity what's the reason for this?

emlaver replied:
RA tests were not running with an error complaining about the simulator iPhone version.  I'll test again to confirm.

emlaver replied:
Here's the `rake testios` output:
```
[CDT_test-IAM-jenkins-ORSF7CTM2EENSTMYD5LWS2L2LFXKBP4FAEOLZSTOZQSMP5P6LKNQ] Running shell script
+ rake testios
xcodebuild: error: Unable to find a destination matching the provided destination specifier:
		{ platform:iOS Simulator, OS:latest, name:iPhone 5 }

	The requested device could not be found because no available devices matched the request.

	Available destinations for the "CDTDatastoreTests" scheme:
		{ platform:iOS, id:c98903079a8fde8a181f1fcfd84b9e4ed84b7d50, name:iPod }
		{ platform:iOS Simulator, id:817CE432-E2BD-47F7-BC03-65D95C9577E7, OS:11.0, name:iPad (5th generation) }
		{ platform:iOS Simulator, id:18E7B059-95B7-40D1-BF91-CE018CA4F8F8, OS:11.0, name:iPad Air }
		{ platform:iOS Simulator, id:0EF5B18A-E22A-4D75-A61B-121E494D09C4, OS:11.0, name:iPad Air 2 }
		{ platform:iOS Simulator, id:285D53A6-F479-41F2-878B-A96D3A792737, OS:11.0, name:iPad Pro (9.7-inch) }
		{ platform:iOS Simulator, id:2FE186DD-617F-4342-AFE1-3EC28810B633, OS:11.0, name:iPad Pro (10.5-inch) }
		{ platform:iOS Simulator, id:CFC00305-150F-44A7-A391-447DAEE9657E, OS:11.0, name:iPad Pro (12.9-inch) }
		{ platform:iOS Simulator, id:FE89E927-58EE-4B82-9B54-5A299D30D179, OS:11.0, name:iPad Pro (12.9-inch) (2nd generation) }
		{ platform:iOS Simulator, id:580B93D6-2C22-47A1-9BC8-026CD6FF6E4D, OS:11.0, name:iPhone 5s }
		{ platform:iOS Simulator, id:B67B75EB-AA50-444C-BA28-53AE4B642CFD, OS:11.0, name:iPhone 6 }
		{ platform:iOS Simulator, id:1760030C-257A-4C5D-AF05-8CF1FFC8F6DB, OS:11.0, name:iPhone 6 Plus }
		{ platform:iOS Simulator, id:45A6EF6C-A522-4901-8FED-8BD483CADFA6, OS:11.0, name:iPhone 6s }
		{ platform:iOS Simulator, id:08A25593-346B-4A30-BD94-7903E2A39FF5, OS:11.0, name:iPhone 6s Plus }
		{ platform:iOS Simulator, id:23D65A6D-DA7B-488A-B709-126660058D01, OS:11.0, name:iPhone 7 }
		{ platform:iOS Simulator, id:6D3C313C-C4FE-4850-81A5-2992E57828CC, OS:11.0, name:iPhone 7 Plus }
		{ platform:iOS Simulator, id:52A6F190-4C24-4D64-AE75-E8C40FF7CF60, OS:11.0, name:iPhone 8 }
		{ platform:iOS Simulator, id:1E595F43-F52F-45A1-8D78-26104A91917B, OS:11.0, name:iPhone 8 Plus }
		{ platform:iOS Simulator, id:4D50B7E8-8821-48E4-9332-B1AAD6F28C3F, OS:11.0, name:iPhone SE }
		{ platform:iOS Simulator, id:1B679759-0684-40F9-A017-F0480EAFD940, OS:11.0, name:iPhone X }

	Ineligible destinations for the "CDTDatastoreTests" scheme:
		{ platform:iOS, id:dvtdevice-DVTiPhonePlaceholder-iphoneos:placeholder, name:Generic iOS Device }
		{ platform:iOS Simulator, id:dvtdevice-DVTiOSDeviceSimulatorPlaceholder-iphonesimulator:placeholder, name:Generic iOS Simulator Device }
rake aborted!
[FAILED] Tests CDTDatastore.xcworkspace, CDTDatastoreTests
/Users/cloudantsync/jenkins-ios/workspace/CDT_test-IAM-jenkins-ORSF7CTM2EENSTMYD5LWS2L2LFXKBP4FAEOLZSTOZQSMP5P6LKNQ/Rakefile:117:in `test'
/Users/cloudantsync/jenkins-ios/workspace/CDT_test-IAM-jenkins-ORSF7CTM2EENSTMYD5LWS2L2LFXKBP4FAEOLZSTOZQSMP5P6LKNQ/Rakefile:67:in `block in <top (required)>'
Tasks: TOP => testios
(See full trace by running task with --trace)
script returned exit code 1
```

ricellis commented on Jenkinsfile:
```diff
@@ -67,6 +67,43 @@ def buildAndTest(nodeLabel, target, rakeEnv, encrypted) {
     }
 }
 
+def buildAndIamTest(nodeLabel, target, rakeEnv, encrypted) {
```
I'd rather this wasn't duplicating most of the code from `buildAndTest` method.
Can't we just surround in an extra `withCredentials` e.g.

```groovy
def buildAndIamTest(nodeLabel, target, rakeEnv, encrypted) {
    withCredentials([string(credentialsId: 'clientlibs-test-iam', variable: 'TEST_COUCH_IAM_API_KEY')]) {
        buildAndTest(nodeLabel, target, rakeEnv, encrypted)
    }
}
```

emlaver replied:
Removed method in 7a2a992.

emlaver commented on CDTDatastoreReplicationAcceptanceTests/Attachments.m:
```diff
@@ -123,131 +110,106 @@ - (void)deleteAttachmentNamed:(NSString*)attachmentName
-    CDTDocumentRevision *rev = [CDTDocumentRevision revisionWithDocId:docId];
-    rev.body = [@{ @"hello" : @12 } mutableCopy];
+    if([self.iamApiKey length] != 0) {
+        NSLog(@"IAM API key enabled, skipping test");
```
I spoke with @tomblench about this testing failing.
His response on Slack:
...`[CDTSavedHTTPAttachment dataFromAttachmentContent]` is trying to download 
the data outside `CDTURLSession` and fails because it doesn’t have the bearer token...
...but i would like to get rid of the whole class because it’s deprecated...
...it was originally intended to support something on worklight/mobilefirst: https://github.com/cloudant/CDTDatastore/blob/33c27b00e577a97420f53807d77218a8f6cc6f04/CDTDatastore/Attachments/CDTAttachment.h#L122-L125

@tomblench should I go ahead and create a PR to remove this class and test?

tomblench replied:
Yes - we might have to check when it was deprecated. If it was deprecated under 0.x then fine, otherwise it will mean a version bump to 2.x (which I don't think is a problem)

emlaver replied:
https://github.com/cloudant/CDTDatastore/pull/236 - the deprecated messages were added in 1.1.0 (2016-06-17).

emlaver replied:
@tomblench @ricellis PR #391 for removal of deprecated `CDTSavedHTTPAttachment `class.

emlaver commented on CDTDatastoreReplicationAcceptanceTests/CloudantReplicationBase.m:
```diff
@@ -35,7 +35,11 @@ - (void)setUp
     if (settings.authorization != nil) {
-        [UNIRest defaultHeader:@"Authorization" value: settings.authorization];
+        if(settings.iamApiKey) {
+           self.iamApiKey = settings.iamApiKey;
```
Fixed in c19d1cb.

emlaver commented:
@tomblench @ricellis I've rebased and removed the duplicated Jenkinsfile code in 7a2a992.  I've still kept the if block commented to run RA tests on this branch.  I will revert that before merging this PR.

emlaver commented:
@tomblench @ricellis I've rebased again to include #390.  Looks like the RA encryption tests are now passing on this branch.

ricellis commented:
Yeah I think this is ok other than the failing test.

emlaver commented:
Opened #392 for failing IAM-enabled RA test.
